src/gtc_unstructured/cpp_util/unstructured/include/gridtools/usid/cuda_helpers.hpp

```diff
@@ -16,37 +17,36 @@
 #include "helpers.hpp"
 
 namespace gridtools::usid::cuda {
-using traits_t = storage::gpu;
-
-inline auto make_allocator() {
-  return sid::device::make_cached_allocator(&cuda_util::cuda_malloc<char[]>);
-}
-
-template <class Kernel, class Ptr, class Strides, class... Neighbors>
-__global__ void kernel(int h_size, Ptr ptr_holder, Strides strides,
-                       Neighbors... neighbors) {
-  auto h = blockIdx.x * blockDim.x + threadIdx.x;
-  if (h >= h_size)
-    return;
-  Kernel()()(
-      sid::shifted(ptr_holder(), device::at_key<dim::h>(strides), h), strides,
-      tuple_util::device::make<pair>(neighbors.first(), neighbors.second)...);
-}
-
-template <class Kernel, class Size, class Sid, class... Sids>
-void call_kernel(Size size, Sid &&fields, Sids &&...neighbor_fields) {
-  int threads_per_block = 32;
-  int blocks = (size + threads_per_block - 1) / threads_per_block;
-  kernel<Kernel><<<blocks, threads_per_block>>>(
-      size, sid::get_origin(fields), sid::get_strides(fields),
-      tuple_util::make<pair>(
-          sid::get_origin(neighbor_fields),
-          at_key<dim::h>(sid::get_strides(neighbor_fields)))...);
-  GT_CUDA_CHECK(cudaGetLastError());
-}
-
-template <class Tag, class Ptr>
-__device__ decltype(auto) field(Ptr const &ptr) {
-  return *device::at_key<Tag>(ptr);
-}
+    using traits_t = storage::gpu;
+
+    inline auto make_allocator() { return sid::device::make_cached_allocator(&cuda_util::cuda_malloc<char[]>); }
+
+    template <class Kernel, class Ptr, class Strides, class... Neighbors>
+    __global__ void kernel(int h_size, int k_size, Ptr ptr_holder, Strides strides, Neighbors... neighbors) {
+        auto h = blockIdx.x * blockDim.x + threadIdx.x;
+        if (h >= h_size)
+            return;
+
+        sid::make_loop<dim::k>(k_size)([&](auto &ptr, auto const &strides) {
```

Comment:
The part that can go to the host is `sid::make_loop<dim::k>(k_size)`, not more right? Then the &-capture is still ok.

Reply:
the only problem would be that the same `tuple_util::device::make<pair>(neighbors.first(), neighbors.second)...` will be executed in a loop and the compiler should be able to optimize that.
Ideally you can do as less as possible here because it is the innermost loop.

Reply:
But it should be on device, right? As I am calling `ptr_holder()`

Reply:
yes